### PR TITLE
Coordinator returns ranges for reading in sorted order

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

I'm not particularly sure about mechanics of the issue (it was occasionally reproduced on a specific fairly large dataset). Maybe it is something around skipping rows of the last granule. It happens only with parallel replicas. Given that I supposed that it might be caused by random order of segments. Now it looks correct, because problem disappeared with this fix.
Anyway ordered segments will at least make debug output less surprising for a reader.

<details>

```
DB::Exception: Invalid version for SerializationLowCardinality key column.: (while reading column type): (while reading from part data/fd2c55b6-f72c-4a25-a9bc-ad5bdf66045a/20240101_25_25_0/ in table default.messages (fd2c55b6-f72c-4a25-a9bc-ad5bdf66045a) located on disk s3WithKeeperDisk of type s3_with_keeper, from mark 275 with max_rows_to_read = 16384): While executing MergeTreeSelect(pool: ReadPoolParallelReplicas, algorithm: Thread). Stack trace:

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c73367b in /usr/bin/clickhouse
1. DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x00000000071cc9ad in /usr/bin/clickhouse
2. DB::SerializationLowCardinality::deserializeBinaryBulkStatePrefix(DB::ISerialization::DeserializeBinaryBulkSettings&, std::shared_ptr<DB::ISerialization::DeserializeBinaryBulkState>&) const @ 0x000000001079f7d0 in /usr/bin/clickhouse
3. DB::MergeTreeReaderCompact::readPrefix(DB::NameAndTypePair const&, std::function<DB::ReadBuffer* (DB::ISerialization::SubstreamPath const&)> const&, std::function<DB::ReadBuffer* (DB::ISerialization::SubstreamPath const&)> const&, std::optional<std::pair<String, unsigned long>> const&) @ 0x00000000124d9794 in /usr/bin/clickhouse
4. DB::MergeTreeReaderCompactMultiBuffer::seekToMarkAndDeserializePrefix(unsigned long, unsigned long, bool) @ 0x00000000124dae3f in /usr/bin/clickhouse
5. DB::MergeTreeReaderCompactMultiBuffer::readRows(unsigned long, unsigned long, bool, unsigned long, std::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn>>>&) @ 0x00000000124db274 in /usr/bin/clickhouse
6. DB::MergeTreeRangeReader::DelayedStream::finalize(std::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn>>>&) @ 0x00000000124c0082 in /usr/bin/clickhouse
7. DB::MergeTreeRangeReader::read(unsigned long, DB::MarkRanges&) @ 0x00000000124c85e0 in /usr/bin/clickhouse
8. DB::MergeTreeRangeReader::read(unsigned long, DB::MarkRanges&) @ 0x00000000124c7e6d in /usr/bin/clickhouse
9. DB::MergeTreeRangeReader::read(unsigned long, DB::MarkRanges&) @ 0x00000000124c7e6d in /usr/bin/clickhouse
10. DB::MergeTreeReadTask::read(DB::MergeTreeReadTask::BlockSizeParams const&) @ 0x00000000124d57f6 in /usr/bin/clickhouse
11. DB::MergeTreeThreadSelectAlgorithm::readFromTask(DB::MergeTreeReadTask&, DB::MergeTreeReadTask::BlockSizeParams const&) @ 0x0000000012e1f28f in /usr/bin/clickhouse
12. DB::MergeTreeSelectProcessor::read() @ 0x00000000124eee30 in /usr/bin/clickhouse
13. DB::MergeTreeSource::tryGenerate() @ 0x0000000012e4bfb8 in /usr/bin/clickhouse
14. DB::ISource::work() @ 0x00000000129bbd43 in /usr/bin/clickhouse
15. DB::ExecutionThreadContext::executeTask() @ 0x00000000129d511a in /usr/bin/clickhouse
16. DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x00000000129cb710 in /usr/bin/clickhouse
17. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x00000000129cc7f8 in /usr/bin/clickhouse
18. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x000000000c81fb81 in /usr/bin/clickhouse
19. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000c8233dc in /usr/bin/clickhouse
20. void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000c8221de in /usr/bin/clickhouse
21. ? @ 0x00007f1d051a3ac3 in ?
22. ? @ 0x00007f1d05235850 in ?
```

</details>

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit b1619218255f75dfa94ee8494757ddfd9795fcf9 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59089/b1619218255f75dfa94ee8494757ddfd9795fcf9/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59089/b1619218255f75dfa94ee8494757ddfd9795fcf9/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 1]
```diff
@@ -56,6 +56,32 @@ takeFromRange(const MarkRange & range, size_t min_number_of_marks, size_t & curr
     current_marks_amount += range_we_take.getNumberOfMarks();
     return range_we_take.getNumberOfMarks();
 }
+
+void sortResponseRanges(RangesInDataPartsDescription & result)
+{
+    std::ranges::sort(result, [](const auto & lhs, const auto & rhs) { return lhs.info < rhs.info; });
```
Comment (antonio2368) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:62]: we have `pdqsort` in `base/base/sort.h` by calling from global namespace `::sort`
do we want to use that?

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp [hunk 2]
```diff
@@ -56,6 +56,32 @@ takeFromRange(const MarkRange & range, size_t min_number_of_marks, size_t & curr
...
+        new_result.back().ranges.insert(
+            new_result.back().ranges.end(),
+            std::make_move_iterator(ranges_in_part.ranges.begin()),
+            std::make_move_iterator(ranges_in_part.ranges.end()));
+        ranges_in_part.ranges.clear();
+    }
+
+    /// Sort ranges for each part
+    for (auto & ranges_in_part : new_result)
+        std::sort(ranges_in_part.ranges.begin(), ranges_in_part.ranges.end());
```
Comment (devcrafter) [inline src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:81]: does it make sense to aggregate ranges here as well?

**Review by devcrafter — APPROVED** [review]

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp (hunk 1 above)
Comment (nickitat) [reply src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:62]: it'd make sense for a perf critical part of the code (like sorting engine), but here I wouldn't try to be too smart

File: src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp (hunk 2 above)
Comment (nickitat) [reply src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp:81]: probably, but the code would be even messier

**Pull request merged**